### ISSUE TITLE
Clarify the sintax for magic comments in ERB files

### ISSUE THREAD

**Issue by afdev82** [glebm/i18n-tasks#435]
Hi all,

I've just upgraded to v1 (1.0.5) and I noticed that the health task now is failing due to some unused translations.
E.g., this piece of code was fine with the v0.9.37:

```erb
<%#-
i18n-tasks-use t('configurations.product_data_sheet_drawings')
i18n-tasks-use t('configurations.shop_drawings')
i18n-tasks-use t('configurations.client_drawings')
i18n-tasks-use t('configurations.production_drawings')
i18n-tasks-use t('configurations.spare_part_glass_drawings')
i18n-tasks-use t('configurations.price_overview_drawings')
i18n-tasks-use t('configurations.wall_template_drawings')
i18n-tasks-use t('configurations.dp_primer_drawings')
i18n-tasks-use t('configurations.dp_primer_black_drawings')
i18n-tasks-use t('configurations.dp_markers_drawings')
i18n-tasks-use t('configurations.dp_sandblasting_drawings')
i18n-tasks-use t('configurations.dp_pattern_drawings') -%>
<%= t "configurations.#{type}_drawings" %>
```

Now if I run `bundle exec i18n-tasks health`, I'm getting unused keys warnings:

```
Unused keys (20) | i18n-tasks v1.0.5
+--------+------------------------------------------+--------------------------------------------------------------+
| Locale | Key                                      | Value                                                        |
+--------+------------------------------------------+--------------------------------------------------------------+
|   de   | configurations.client_drawings           | Fertigungszeichnungen (Client Drawings)                      |
|   de   | configurations.dp_markers_drawings       | Digitaldrucker-Markierungen-Zeichnungen                      |
|   de   | configurations.dp_primer_black_drawings  | Digitaldrucker-Schwarz-Grundierung-Zeichnungen               |
|   de   | configurations.dp_primer_drawings        | Digitaldrucker-Grundierung-Zeichnungen                       |
|   de   | configurations.dp_sandblasting_drawings  | Digitaldrucker-Sandblasting-Zeichnungen                      |
|   de   | configurations.price_overview_drawings   | Preisübersicht                                               |
|   de   | configurations.production_drawings       | Produktionszeichnungen (Production Drawings)                 |
|   de   | configurations.shop_drawings             | Fertigungszeichnungen (Shop Drawings)                        |
|   de   | configurations.spare_part_glass_drawings | Zeichnungen der Glas Ersatzteile (Spare Part Glass Drawings) |
|   de   | configurations.wall_template_drawings    | Wandvorlage-Zeichnungen                                      |
|   en   | configurations.client_drawings           | Client Drawings                                              |
|   en   | configurations.dp_markers_drawings       | Digital Printer Markers Drawings                             |
|   en   | configurations.dp_primer_black_drawings  | Digital Printer Primer Black Drawings                        |
|   en   | configurations.dp_primer_drawings        | Digital Printer Primer Drawings                              |
|   en   | configurations.dp_sandblasting_drawings  | Digital Printer Sandblasting Drawings                        |
|   en   | configurations.price_overview_drawings   | Price Overview                                               |
|   en   | configurations.production_drawings       | Production Drawings                                          |
|   en   | configurations.shop_drawings             | Shop Drawings                                                |
|   en   | configurations.spare_part_glass_drawings | Spare Part Glass Drawings                                    |
|   en   | configurations.wall_template_drawings    | Wall template Drawings                                       |
```
I noticed that if I write the code in different ways I'm getting different results:

```erb
<%# i18n-tasks-use t('configurations.product_data_sheet_drawings')
i18n-tasks-use t('configurations.shop_drawings')
i18n-tasks-use t('configurations.client_drawings')
i18n-tasks-use t('configurations.production_drawings')
i18n-tasks-use t('configurations.spare_part_glass_drawings')
i18n-tasks-use t('configurations.price_overview_drawings')
i18n-tasks-use t('configurations.wall_template_drawings')
i18n-tasks-use t('configurations.dp_primer_drawings')
i18n-tasks-use t('configurations.dp_primer_black_drawings')
i18n-tasks-use t('configurations.dp_markers_drawings')
i18n-tasks-use t('configurations.dp_sandblasting_drawings')
i18n-tasks-use t('configurations.dp_pattern_drawings') %>
<%= t "configurations.#{type}_drawings" %>
```
```
$ bundle exec i18n-tasks health
Unused keys (2) | i18n-tasks v1.0.5
+--------+------------------------------------+-----------------------------------+
| Locale | Key                                | Value                             |
+--------+------------------------------------+-----------------------------------+
|   de   | configurations.dp_pattern_drawings | Digitaldrucker-Muster-Zeichnungen |
|   en   | configurations.dp_pattern_drawings | Digital Printer Pattern Drawings  |
+--------+------------------------------------+-----------------------------------+
```
In this way I'm still getting a warning for the last entry, but if I write the comments inside normal <% %> tags, it's fine:

```erb
<% # i18n-tasks-use t('configurations.product_data_sheet_drawings')
# i18n-tasks-use t('configurations.shop_drawings')
# i18n-tasks-use t('configurations.client_drawings')
# i18n-tasks-use t('configurations.production_drawings')
# i18n-tasks-use t('configurations.spare_part_glass_drawings')
# i18n-tasks-use t('configurations.price_overview_drawings')
# i18n-tasks-use t('configurations.wall_template_drawings')
# i18n-tasks-use t('configurations.dp_primer_drawings')
# i18n-tasks-use t('configurations.dp_primer_black_drawings')
# i18n-tasks-use t('configurations.dp_markers_drawings')
# i18n-tasks-use t('configurations.dp_sandblasting_drawings')
# i18n-tasks-use t('configurations.dp_pattern_drawings') %>
<%= t "configurations.#{type}_drawings" %>
```
```
$ bundle exec i18n-tasks health
✓ Good job! Every translation is in use.
```

Well, so what is the right way to use the magic comments inside the ERB files?
Thank you for your support!

**Comment by afdev82** [comment]
If I use the last syntax, then the code gets parsed by [erb-lint](https://github.com/Shopify/erb-lint) and I get errors there about wrong indentation, so I'm in a loop :see_no_evil:

**Comment by afdev82** [comment]
I've just found that if I write in this way

```erb
<%# i18n-tasks-use t('configurations.product_data_sheet_drawings')
# i18n-tasks-use t('configurations.shop_drawings')
# i18n-tasks-use t('configurations.client_drawings')
# i18n-tasks-use t('configurations.production_drawings')
# i18n-tasks-use t('configurations.spare_part_glass_drawings')
# i18n-tasks-use t('configurations.price_overview_drawings')
# i18n-tasks-use t('configurations.wall_template_drawings')
# i18n-tasks-use t('configurations.dp_primer_drawings')
# i18n-tasks-use t('configurations.dp_primer_black_drawings')
# i18n-tasks-use t('configurations.dp_markers_drawings')
# i18n-tasks-use t('configurations.dp_sandblasting_drawings')
# i18n-tasks-use t('configurations.dp_pattern_drawings') -%>
<%= t "configurations.#{type}_drawings" %>
```
changing the closing tag to `-%>` and use # for all magic comments, it's working.
But it seems more an hack than a solution.

**Comment by davidwessman** [comment]
> @davidwessman Seems like we might need to do a bit more AST hacking, can you please have a look at this and the other issue?

Yes, we need to decide on a format!

**Comment by glebm** [comment]
Ideally we'd support all of the above

**Comment by davidwessman** [comment]
I have started working on handling the ERB-comments differently that should help for both this issue and #434 but did not manage to get it completely working yet.

**Comment by davidwessman** [comment]
@glebm Should all of these formats be supported?
I am thinking that I will make the erb-ast-parser handle as many of these cases as possible, maybe the first cases with ruby-comments inside one erb-tag can be handled by the normal comments-matching?

```erb
<!-- Inside the same ERB-tag -->
<%
  # i18n-tasks-use t('ruby_comment_same_erb_tag')
  inside_erb = FakeTranslate.single_inside
%>

<%
  # i18n-tasks-use t('multiple_ruby_comment_same_erb_tag1')
  # i18n-tasks-use t('multiple_ruby_comment_same_erb_tag2')
  # i18n-tasks-use t('multiple_ruby_comment_same_erb_tag3')
  inside_erb = FakeTranslate.multiple_inside
%>


<!-- ERB-comments before ERB-tag -->
<%# i18n-tasks-use t('erb_comment_before_erb_tag') %>
<%= FakeTranslate.with_erb_comment_before %>

<%# i18n-tasks-use t('multiple_erb_comment_before_erb_tag1') %>
<%# i18n-tasks-use t('multiple_erb_comment_before_erb_tag2') %>
<%# i18n-tasks-use t('multiple_erb_comment_before_erb_tag3') %>
<%= FakeTranslate.multiple_with_erb_comment_before %>

<%# i18n-tasks-use t('multiple_erb_comment_before_erb_tag1')
i18n-tasks-use t('multiple_erb_comment_before_erb_tag2')
i18n-tasks-use t('multiple_erb_comment_before_erb_tag3') %>
<%= FakeTranslate.multiple_one_tag_with_erb_comment_before %>


<!-- Ruby-comments before ERB-tag -->
<% # i18n-tasks-use t('ruby_comment_before_erb_tag') %>
<%= FakeTranslate.with_ruby_comment_before %>

<% # i18n-tasks-use t('multiple_ruby_comment_before_erb_tag1') %>
<% # i18n-tasks-use t('multiple_ruby_comment_before_erb_tag2') %>
<% # i18n-tasks-use t('multiple_ruby_comment_before_erb_tag3') %>
<%= FakeTranslate.multiple_with_ruby_comment_before %>

<%
  # i18n-tasks-use t('multiple_ruby_comment_before_erb_tag1')
  # i18n-tasks-use t('multiple_ruby_comment_before_erb_tag2')
  # i18n-tasks-use t('multiple_ruby_comment_before_erb_tag3')
%>
<%= FakeTranslate.multiple_one_tag_with_ruby_comment_before %>

```

**Comment by glebm** [comment]
Yeah, I think all of the above are variations of how people write ERB comments, so they should ideally all be supported

**Comment by davidwessman** [comment]
> Yeah, I think all of the above are variations of how people write ERB comments, so they should ideally all be supported

I got very confused when trying to fix both #434 and this one at the same time, but will try to tackle this one when #437 is merged.

**Comment by davidwessman** [comment]
@glebm I think #437 solves this issue as well, I added some more tests in #440 and it seems to work okay.

**Comment by davidwessman** [comment]
@afdev82 could you try to install from the main-branch and see if it works better for you now?

**Comment by davidwessman** [comment]
@glebm Can we release a new version with these fixes?

@afdev82 You cannot mix ERB-comments (`<%# %>` ) with ruby-comments (`# i18n-tasks-use t('what')`).

So your examples would only work if you add a space before `<%#` -> `<% #` or remove the surplus `#` on each row inside `<%# %>`

**Comment by afdev82** [comment]
OK, I didn't know that.
Then it's all fine!

**Comment by davidwessman** [comment]
> OK, I didn't know that.
> Then it's all fine!

Oh, no worries! I had to find out the difference for this issue.

**Comment by glebm** [comment]
Released v1.0.6!
